src/dbus/common/dbus_message_helper.hpp

```diff
@@ -182,20 +251,19 @@ template <typename T> otbrError DBusMessageExtract(DBusMessageIter *aIter, T &aV
 
 template <typename T> otbrError DBusMessageExtract(DBusMessageIter *aIter, std::vector<T> &aValue)
 {
-    otbrError err = OTBR_ERROR_NONE;
-    uint32_t  size;
+    otbrError       err = OTBR_ERROR_NONE;
+    DBusMessageIter subIter;
 
-    VerifyOrExit(dbus_message_iter_get_arg_type(aIter) == DBUS_TYPE_UINT32, err = OTBR_ERROR_DBUS);
-    SuccessOrExit(err = DBusMessageExtract(aIter, size));
+    VerifyOrExit(dbus_message_iter_get_arg_type(aIter) == DBUS_TYPE_ARRAY, err = OTBR_ERROR_DBUS);
+    dbus_message_iter_recurse(aIter, &subIter);
 
     aValue.clear();
-    for (uint32_t i = 0; i < size; i++)
+    while (dbus_message_iter_get_arg_type(&subIter) != DBUS_TYPE_INVALID)
     {
         aValue.emplace_back();
-        DBusMessageExtract(aIter, aValue.back());
+        SuccessOrExit(err = DBusMessageExtract(&subIter, aValue.back()));
```

Comment:
We never handle partial data, if you get an error code you should abort the parse.

Reply:
**We should keep the output `aValue` as it was before calling this function when this function failed**. Because it makes no sense to the caller that a failed function calling get its data modified, especially, this function will emplace an uninitialized data at the back. Suppose the user is calling this function on its member data, like this:

```C++
otbrError error = OTBR_ERROR_NONE;
SuccessOrExit(error = GetChannelMonitorAllChannelQualities(&aIter, mChannelQualities));

exit:

```

This calling and handling is common but leaves `mChannelQualities` in an invalid state. We can of cause use a temp variable for this case, but that will be tedious:

```C++
otbrError error = OTBR_ERROR_NONE;
ChannelQualityArray qualities;
SuccessOrExit(error = GetChannelMonitorAllChannelQualities(&aIter, qualities));
mChannelQualities = qualities;

exit:
```

What do you think?

Reply:
There is a PR about these error encoding issues: https://github.com/openthread/ot-br-posix/pull/373

Reply:
Fixed